### ISSUE TITLE
plugin/forward: prewarm connections

### ISSUE THREAD

**Issue by miekg** [coredns/coredns#2733]
It seems sensible to prewarm the persistent connections cache in the forward plugin. This (hopefully) prevents a thundering herd problem when a lot of request come in and they all want to be served from a TLS connection.

A couple of choices need to be made here:
* Does it make sense for all transports? Or just tcp and tcp-tls? Or only tls? Or even udp?
* Should we be smart and check the length of the remaining connections? We can be pretty dump and just call this on a time.Ticker and do it every second
* We clear the map with a `clear` function; if we keep this connection cache to be always filled, this code does not make sense anymore.

**Comment by WinstonPrivacy** [comment]
I've done a fair amount of work along these lines with ordinary TLS connections (HTTPS, not DNS though I expect them to follow the same principles).

The expensive part of this is the handshake and round trip which occurs during the initial negotiation of the TLS tunnel. On low end ARM64 boards, a large part of CPU utilization can be tied up here. To the extent that you can maintain a connection pool of open TLS tunnels, CPU utilization and responsiveness can be dramatically enhanced. 

The benefits of this will primarily be realized for DNS-over-TLS. TLS handshakes are independent of the underlying connection protocol (TCP/UDP). 

Golang's net/http Transport implements TLS connection pooling, so this may serve as a relevant example. One key learning taken from there is to set a sensible upper limit on the number of pooled connections.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>I've done a fair amount of work along these lines with ordinary TLS connections (HTTPS, not DNS though I expect them to follow the same principles).
>
>The expensive part of this is the handshake and round trip which occurs during the initial negotiation of the TLS tunnel. On low end ARM64 boards, a large part of CPU utilization can be tied up here. To the extent that you can maintain a connection pool of open TLS tunnels, CPU utilization and responsiveness can be dramatically enhanced.
>
>The benefits of this will primarily be realized for DNS-over-TLS. TLS handshakes are independent of the underlying connection protocol (TCP/UDP).
>
>Golang's net/http Transport implements TLS connection pooling, so this may serve as a relevant example. One key learning taken from there is to set a sensible upper limit on the number of pooled connections.

Thanks. Yeah, I wondered if we could just steal what is happening in net/http 
and use that.

Also gating the number of concurrent *new* connections to GOMAXPROCS (for TLS at 
least), might make sense here.

**Comment by WinstonPrivacy** [comment]
The net/http code is kind of hard to follow and it's pretty intricately tied into HTTP specifics. However, the general principles are straightforward:

- The connection pool consists of a map of persistent connections.
- The user can specify if connection pooling should be used at all and how long idle connections should be kept alive for.
- Persistent connections are opened as needed
- A persistent connection (persistConn) wraps a Transport with various helpers needed to maintain the pool.
- persistConns are mapped by protocol, host and port. For instance, "tls:9.9.9.9:853" or "dns:4.4.4.4:53".
- There are user configurable limits to the number of idle connections allowed for each key (host/protocol/port) as well as in total.
- Private helper functions are provided to get and put idle connections, as well as to close all open idle connections.
- Finally, a single function (RoundTrip() in the standard HTTP library) is provided for callers to conduct a one-shot request to the destination. This function takes care of all handshaking and pooling, abstracting the details away from the user.

With the above in mind, following the standard transport:RoundTrip() logic should be a little easier.

**Comment by miekg** [comment]
Thanks. I've tried looking at the code before, but the connection pooling
got hairy quickly

**Comment by WinstonPrivacy** [comment]
Yes, it is overly convoluted, though the relevant parts can be easily
extracted (and refactored if necessary).

**Comment by WinstonPrivacy** [comment]
Possibly helpful: https://tools.ietf.org/id/draft-ietf-dprive-dns-over-tls-05.html

> For DNS clients that use library functions such as "getaddrinfo()" and "gethostbyname()", current implementations are known to open and close TCP connections each DNS call. **To avoid excess TCP connections, each with a single query, clients SHOULD reuse a single TCP connection to the recursive resolver.** Alternatively they may prefer to use UDP to a DNS-over-TLS enabled caching resolver on the same machine that then uses a system-wide TCP connection to the recursive resolver.
> 
> **In order to amortize TCP and TLS connection setup costs, clients and servers SHOULD NOT immediately close a connection after each response. Instead, clients and servers SHOULD reuse existing connections for subsequent queries as long as they have sufficient resources. In some cases, this means that clients and servers may need to keep idle connections open for some amount of time.**
> 
> Proper management of established and idle connections is important to the healthy operation of a DNS server. An implementor of DNS-over-TLS SHOULD follow best practices for DNS-over-TCP, as described in [I-D.ietf-dnsop-5966bis]. Failure to do so may lead to resource exhaustion and denial-of-service.
> 
> Whereas client and server implementations from the [RFC1035] era are known to have poor TCP connection management, this document stipulates that successful negotiation of TLS indicates the willingness of both parties to keep idle DNS connections open, independent of timeouts or other recommendations for DNS-over-TCP without TLS. In other words, software implementing this protocol is assumed to support idle, persistent connections and be prepared to manage multiple, potentially long-lived TCP connections.
> 
> This document does not make specific recommendations for timeout values on idle connections. Clients and servers should reuse and/or close connections depending on the level of available resources. Timeouts may be longer during periods of low activity and shorter during periods of high activity. Current work in this area may also assist DNS-over-TLS clients and servers select useful timeout values [I-D.edns-tcp-keepalive] [tdns].
> 
> Clients and servers that keep idle connections open MUST be robust to termination of idle connection by either party. As with current DNS-over-TCP, DNS servers MAY close the connection at any time (perhaps due to resource constraints). As with current DNS-over-TCP, clients MUST handle abrupt closes and be prepared to reestablish connections and/or retry queries.
> 
> When reestablishing a DNS-over-TCP connection that was terminated, as discussed in [I-D.ietf-dnsop-5966bis], TCP Fast Open [RFC7413] is of benefit. Underlining the requirement for sending only encrypted DNS data on a DNS-over-TLS port (Section 3.2), when using TCP Fast Open the client and server MUST immediately initiate or resume a TLS handshake (clear text DNS MUST NOT be exchanged). DNS servers SHOULD enable fast TLS session resumption [RFC5077] and this SHOULD be used when reestablishing connections.
> 
> When closing a connection, DNS servers SHOULD use the TLS close-notify request to shift TCP TIME-WAIT state to the clients. Additional requirements and guidance for optimizing DNS-over-TCP are provided by [RFC5966], [I-D.ietf-dnsop-5966bis].

**Comment by WinstonPrivacy** [comment]
I was just looking into this some more and I'm a little curious. I see persistConn in the Forward plugin, in persistent.go. But I can't see it being used anywhere. Is there a chance this is already (even partially) implemented?

**Comment by WinstonPrivacy** [comment]
Nevermind. I think I see it now.

```
func (t *Transport) Dial(proto string) (*dns.Conn, bool, error) {
	// If tls has been configured; use it.
	if t.tlsConfig != nil {
		proto = "tcp-tls"
	}

	t.dial <- proto
c := <-t.ret
```

I thought I understood the problem, but now I'm not sure I do. It seems that a pool of TLS connections is already being maintained. So couldn't you just wrap this section of code with a check to ensure that no more than X simultaneous connections are already being initiated and return a "GoAwayTooBusy" error if this is exceeded? It seems like that along with enforcing a 5-second timeout on the Dial() should prevent the herding error.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/forwar..." ]
>Nevermind. I think I see it now.
>
>`func (t *Transport) Dial(proto string) (*dns.Conn, bool, error) {
>	// If tls has been configured; use it.
>	if t.tlsConfig != nil {
>		proto = "tcp-tls"
>	}
>
>	t.dial <- proto
>c := <-t.ret`
>
>I thought I understood the problem, but now I'm not sure I do. It seems that a pool of TLS connections is already being maintained. So couldn't you just wrap this section of code with a check to ensure that no more than X simultaneous connections are already being initiated and return a "GoAwayTooBusy" error if this is exceeded? It seems like that along with enforcing a 5-second timeout on the Dial() should prevent the herding error.

...maybe... sounds like that could work. Could you try it?
Another thing that's worth looking into is if we need to make this a tls.Conn or 
something? I.e. are we re-negotiating TLS on every transmit?

**Comment by WinstonPrivacy** [comment]
I'll take a look. IIRC connections were being dialed through the TLS
library which would mean that they are not handshaking on every request.

**Comment by miekg** [comment]
Ping @WinstonPrivacy are working on any changes in this area?

**Comment by WinstonPrivacy** [comment]
We haven't done anything recently but can report that connection pooling works much better (ie: drastically lower CPU utilization) if you set expire to 30s and our new flag, max_requests to 50.

**Comment by miekg** [comment]
Ok. Did you send a PR for that new flag? Would be good to merge/have a
discussion on that

**Comment by miekg** [comment]
hard to do and not sure this will make a difference.
